File: Intro to Algorithms/PA3/Scripts/List.c
//Galen Turoci
//gturoci
//pa4

#include<stdio.h>
#include<stdlib.h>
#include "List.h"

//structs --------------------------------------------------------------------

//private NodeObj type
typedef struct NodeObj {
	int data;
	struct NodeObj* next;
	struct NodeObj* prev;
} NodeObj;

//private Node type
typedef NodeObj* Node;

//private ListObj type
typedef struct ListObj {
	Node front;
	Node back;
	Node cursor;
	int length;
	int index;
} ListObj;

//Constructors-Destructors ---------------------------------------------------

//newNode()
//Returns reference to new Node object. Initializes next and data fields.
//Private.
Node newNode(int data) {
	Node N = malloc(sizeof(NodeObj));
	N->data = data;
	N->next = NULL;
	N->prev = NULL;
	return(N);
}

//freeNode()
//Frees heap memory pointed to by *pN, sets *pN to NULL.
//Private.
void freeNode(Node* pN) {
	if (pN != NULL && *pN != NULL) {
		free(*pN);
		*pN = NULL;
	}
}

//newList()
//Returns reference to new empty List object.
List newList(void) {
	List L;
	L = malloc(sizeof(ListObj));
	L->front = L->back = L->cursor = NULL;
	L->length = 0;
	L->index = -1;
	return(L);
}


//freeList()
//Frees all heap memory associated with List *pL, and sets *pL to NULL.
void freeList(List* pL) {
	if (pL != NULL && *pL != NULL) { 
		while (length(*pL) > 0)
			deleteFront(*pL);
		free(*pL);
		*pL = NULL;
	}
}

// Access functions -----------------------------------------------------------

//length()
//Returns the length of L.
int length(List L) {
	if (L == NULL) {
		printf("List Error: calling length() on NULL List reference\n");
		exit(1);
	}
	return (L->length);
}

//index()
//Returns the index of L
int index(List L) {
	if (L == NULL) {
      printf("List Error: calling index() on NULL List reference\n");
      exit(1);
   }
   return (L->index);
}

//front()
//Returns the value at the front of L.
//Pre: length(L) < 1
int front(List L) {
	if (L == NULL) {
		printf("List Error: calling front() on NULL List reference\n");
		exit(1);
	} if (length(L) < 1) {
		printf("List Error: calling front() on an empty List\n");
		exit(1);
	}
	return(L->front->data);
}

//back()
//returns the value at the back of this List
//Pre: length(L) > 0
int back(List L) {
	if (L == NULL) {
		printf("List Error: calling back() on NULL List reference\n");
		exit(1);
	}
	if (length(L) < 1) {
		printf("List Error: calling back() on an empty List\n");
		exit(1);
	}
	return(L->back->data);
}

//get()
//Returns cursor element
//Pre: length(L)>0, index(L)>=0
int get(List L) {
	if (L == NULL) {
		printf("List Error: calling front() on NULL List reference\n");
		exit(1);
	}
	if (length(L) < 1) {
		printf("List Error: calling front() on an empty List\n");
		exit(1);
	}
	return(L->cursor->data);
}

//equals()
//returns true (1) if A is identical to B, false (0) otherwise
int equals(List A, List B) {
	int eq = 0;
	Node N = NULL;
	Node M = NULL;

	if (A == NULL || B == NULL) {
		printf("List Error: calling equals() on NULL List reference\n");
		exit(1);
	}

	eq = (length(A) == length(B));
	N = A->front;
	M = B->front;
	while (eq && N != NULL) {
		eq = (N->data == M->data);
		N = N->next;
		M = M->next;
	}
	return eq;
}

//Manipulation procedures ----------------------------------------------------

//clear()
//Resets this List to its original empty state
void clear(List L) {
	if (L == NULL) {
		printf("List Error: calling clear() on NULL List reference\n");
		exit(1);
	}
	while (length(L) > 0)
		deleteFront(L);
}

//moveFront()
//If List is non-empty, places the cursor under the front element
//Does nothing otherwise
void moveFront(List L) {
	if (L == NULL) {
		printf("List Error: calling moveFront() on NULL List reference\n");
		exit(1);
	} if (length(L) > 0) {
		L->cursor = L->front;
		L->index = 0;
	}
}

//moveBack()
//If List is non-empty, places the cursor under the back element
//Does nothing otherwise
void moveBack(List L) {
	if (L == NULL) {
		printf("List Error: calling moveFront() on NULL List reference\n");
		exit(1);
	}
	if (length(L) > 0) {
		L->cursor = L->back;
		L->index = L->length - 1;
	}
}

//movePrev()
//If cursor is defined and not at front, moves cursor one step toward
//front of this List, if cursor is defined and at front, cursor becomes
//undefined, if cursor is undefined does nothing.
void movePrev(List L) {
	if (L == NULL) {
		printf("List Error: calling moveFront() on NULL List reference\n");
		exit(1);
	}
	if (L->cursor == L->front)
		L->cursor = NULL;
	else
		L->cursor = L->cursor->prev;
	L->index--;
}

//moveNext()
//If cursor is defined and not at back, moves cursor one step toward
//back of this List, if cursor is defined and at back, cursor becomes
//undefined, if cursor is undefined does nothing.
void moveNext(List L) {
	if (L == NULL) {
		printf("List Error: calling moveFront() on NULL List reference\n");
		exit(1);
	}
	if (L->cursor == L->back) {
		L->cursor = NULL;
		L->index = -1;
	}
	else {
		L->cursor = L->cursor->next;
		L->index++;
	}
}

//prepend()
//Places new data element at the front of L
void prepend(List L, int data) {
	if (L == NULL) {
		printf("List Error: calling append() on NULL List reference\n");
		exit(1);
	}
	
	Node N = newNode(data);
	if (length(L) < 1)
		L->front = L->back = N; 
	else { 
		L->front->prev = N;
		N->next = L->front;
		L->front = N;
	}
	if (L->cursor != NULL)
		L->index++;
	L->length++;
}

//append()
//Places new data element at the end of L
void append(List L, int data) {
	if (L == NULL) {
		printf("List Error: calling append() on NULL List reference\n");
		exit(1);
	}
	
	Node N = newNode(data);
	if (length(L) < 1) 
		L->front = L->back = N; 
	else { 
		L->back->next = N;
		N->prev = L->back;
		L->back = N; 
	}
	L->length++;
}

//insertBefore()
//Insert a new element before the cursor
//Pre: length(L) > 0, index(L) >= 0
void insertBefore(List L, int data) {
	if (L == NULL) {
		printf("List Error: calling deleteFront() on NULL List reference\n");
		exit(1);
	}
	if (length(L) < 1 || index(L) < 0) {
		printf("List Error: calling deleteFront() on an empty List\n");
		exit(1);
	}
	else if (L->cursor == L->front)
		prepend(L, data);
	else {
		Node N = newNode(data);
		Node M = L->cursor->prev;
		L->cursor->prev = N;
		N->prev = M;
		M->next = N;
		N->next = L->cursor; 
		L->length++;
		L->index++;
	}
}

//insertAfter()
//Insert a new element after the cursor
//Pre: length(L) > 0, index(L) >= 0
void insertAfter(List L, int data) {
	if (L == NULL) {
		printf("List Error: calling deleteFront() on NULL List reference\n");
		exit(1);
	}
	if (length(L) < 1 || index(L) < 0) {
		printf("List Error: calling deleteFront() on an empty List\n");
		exit(1);
	}
	else if (L->cursor == L->back)
		append(L, data);
	else {
		Node N = newNode(data);
		Node M = L->cursor->next;
		L->cursor->next = N;
		N->next = M;
		M->prev = N;
		N->prev = L->cursor; 
		L->length++;
	}
}

//deleteFront()
//Deletes element at front of L
//Pre: length(L) > 0
void deleteFront(List L) {
	if (L == NULL) {
		printf("List Error: calling deleteFront() on NULL List reference\n");
		exit(1);
	}
	if (length(L) < 1) {
		printf("List Error: calling deleteFront() on an empty List\n");
		exit(1);
	}
	
	Node N = NULL;
	N = L->front;
	if (length(L) == 1) {
		L->front = L->back = L->cursor = NULL;
		L->index = -1;
	} else {
		L->front = L->front->next; 
		L->index--;	
	}
	L->length--;
	freeNode(&N);
}

//deleteBack()
//Deletes element at back of L
//Pre: length(L) > 0
void deleteBack(List L) {
	if (L == NULL) {
		printf("List Error: calling deleteFront() on NULL List reference\n");
		exit(1);
	}
	if (length(L) < 1) {
		printf("List Error: calling deleteFront() on an empty List\n");
		exit(1);
	}
	
	Node N = NULL;
	N = L->back;
	if (length(L) == 1) {  
		L->front = L->back = L->cursor = NULL;
		L->index = -1;
	} else {
		if (L->cursor == L->back) {
			L->cursor = NULL;
			L->index = -1;
		}
		L->back = L->back->prev;
	}
	L->length--;
	freeNode(&N);
}

//delete()
//Deletes the cursor element, making cursor undefined
//Pre: length(L) > 0, index(L) >= 0
void delete(List L) {
	if (L == NULL) {
		printf("List Error: calling deleteFront() on NULL List reference\n");
		exit(1);
	}
	if (length(L) < 1 || index(L) < 0) {
		printf("List Error: calling deleteFront() on an empty List\n");
		exit(1);
	}
	
	if (L->cursor == L->back)
		deleteBack(L);
	else if (L->cursor == L->front)
		deleteFront(L);
	else {
		Node N = NULL;
		N = L->cursor;
		Node M = L->cursor->next;
		L->cursor = L->cursor->prev;
		L->cursor->next = M;
		L->index = -1;
		L->length--;
		freeNode(&N);
	}
}

//Other Functions ------------------------------------------------------------

//printList()
//Prints data elements in L on a single line to stdout.
void printList(FILE* out, List L) {
	Node N = NULL;

	if (L == NULL) {
		printf("List Error: calling printList() on NULL List reference\n");
		exit(1);
	}

	for(N = L->front; N != NULL; N = N->next)
		fprintf(out, "%d ", N->data);
	fprintf(out, "\n");
}

//copyList()
//Copies the List
List copyList(List L) {
	List LC = newList();
	Node N = L->front;

	while (N != NULL) {
		append(LC, N->data);
		N = N->next;
	}
	return LC;
}

File: Intro to Algorithms/PA2/Reference Scripts/List.java
//Galen Turoci
//gturoci
//pa3

class List {

	private class Node {
		// Fields
		Object data;
		Node next;
		Node prev;
      
		// Constructor
		Node(Object data) { 
		this.data = data;
		next = null; 
		prev = null;
		}
      
		// toString():  overrides Object's toString() method
		public String toString() { 
			return String.valueOf(data); 
		}
      
		// equals(): overrides Object's equals() method
		public boolean equals(Object x) {
			boolean eq = false;
			Node that;
			if (x instanceof Node) {
				that = (Node) x;
				eq = ((this.data).equals(that.data));
			}
			return eq;
		}
	}

	// Fields
	private Node front;
	private Node back;
	private Node cursor;
	private int length;
	private int index;

	// Constructor ------------------------------------------------------------
	List() { 
		front = back = cursor = null; 
		length = 0;
		index = -1;
	}


	//Access Functions --------------------------------------------------------

	//length()
	//Returns the number of elements in this List.
	int length() { 
		return length; 
	}
	
	//index()
	//Returns the index of the cursor element if it is defined
	//returns -1 otherwise
	int index() {
		return index;
	}

	//front() 
	//Returns front element.
	//Pre: length()>0
	Object front() {
		if (this.length <= 0)
			throw new RuntimeException("List Error: front() called on empty List");
		return front.data;	
	}
	
	//back()
	//Returns back element
	//Pre: length()>0
	Object back() {
		if (this.length <= 0)
			throw new RuntimeException("List Error: back() called on empty List");
		return back.data;
	}
	
	//get()
	//Returns cursor element
	//Pre: length()>0, index()>=0
	Object get() {
		if (this.length <= 0 || this.index() < 0)
			throw new RuntimeException("List Error: get() called on an empty List");
		if (cursor == null)
			throw new RuntimeException("List Error: get() called on a null cursor");
		return cursor.data;
	}
	
	//equals(Object x)
	//Returns true if and only if this Object and x are the same
	//integer sequence. The states of the cursors in the two Lists
	//are not used in determining equality
	//overrides Object's equals() method
	public boolean equals(Object x) {
		boolean eq  = false;
		List that;
		Node Y, Z;
		if (x instanceof List) {
			that = (List) x;
			eq = (this.length == that.length);
			Y = this.front;
			Z = that.front;
			while (eq && Y != null) {
				eq = (Y.equals(Z));
				Y = Y.next;
				Z = Z.next;
			}
		}
		return eq;
	}

	// Manipulation Procedures -------------------------------------------------
	
	//clear()
	//Resets this List to its original empty state
	void clear() {
		while (this.length > 0)
			deleteFront();
	}
	
	//moveFront()
	//If List is non-empty, places the cursor under the front element
	//Does nothing otherwise
	void moveFront() {
		if (this.length > 0) {
			cursor = front;
			index = 0;
		}
	}
	
	//moveBack()
	//If List is non-empty, places the cursor under the back element,
	//otherwise does nothing.
	void moveBack() {
		if (this.length > 0) {
			cursor = back;
			index = length - 1;
		}
	}
	
	//movePrev()
	//If cursor is defined and not at front, moves cursor one step toward
	//front of this List, if cursor is defined and at front, cursor becomes
	//undefined, if cursor is undefined does nothing.
	void movePrev() {
		if (cursor == front)
			cursor = null;
		else
			cursor = cursor.prev;
		index--;
	}
	
	//moveNext()
	//If cursor is defined and not at back, moves cursor one step toward
	//back of this List, if cursor is defined and at back, cursor becomes
	//undefined, if cursor is undefined does nothing.
	void moveNext() {
		if (cursor == back) {
			cursor = null;
			index = -1;
		}
		else {
			cursor = cursor.next;
			index++;
		}
	}
	
	//prepend(Object data)
	//Prepends data to the front of this List.
	void prepend(Object data) {
		Node N = new Node(data);
		if (this.length == 0)
			front = back = N;
		else { 
			front.prev = N;
			N.next = front;
			front = N;
		}
		if (cursor != null)
			index++;
		length++;
	}

	//append(Object data)
	//Appends data to back of this List.
	void append(Object data) {
		Node N = new Node(data);
		if (this.length == 0)
			front = back = N;
		else { 
			back.next = N;
			N.prev = back;
			back = N; 
		}
		length++;
	}
	
	//insertBefore(Object data)
	//Insert a new element before the cursor
	//Pre: length()>0, index()>=0
	void insertBefore(Object data) {
		if (this.length <= 0 || this.index() < 0)
			throw new RuntimeException("List Error: insertBefore() called on an empty List");
		else if (cursor == front)
			prepend(data);
		else {
			Node N = new Node(data);
			Node M = cursor.prev;
			cursor.prev = N;
			N.prev = M;
			M.next = N;
			N.next = cursor;
			length++;
			index++;
		}
	}
	
	//insertAfter(Object data)
	//Inster a new element after the cursor
	//Pre: length()>0, index()>=0
	void insertAfter(Object data) {
		if (this.length <= 0 || this.index() < 0)
			throw new RuntimeException("List Error: insertAfter() called on an empty List");
		else if (cursor == back)
			append(data);
		else {
			Node N = new Node(data);
			Node M = cursor.next;
			cursor.next = N;
			N.next = M;
			M.prev = N;
			N.prev = cursor; 
			length++;
		}
	}
	
	//deleteFront()
	//Deletes front element from this List.
	//Pre: length()>0
	void deleteFront() {
		if (this.length <= 0)
			throw new RuntimeException("List Error: deleteFront() called on empty List");
		if (this.length == 1) {
			front = back = cursor = null;
			index = -1;
		} else {
			front = front.next;
			index--;
		}
		length--;
	}
	
	//deleteBack()
	//Deletes the back element from this List
	//Pre: length()>0
	void deleteBack(){
		if (this.length <= 0)
			throw new RuntimeException("List Error: deleteBack() called on empty List");
		if (this.length == 1) {
			back = front = cursor = null;
			index = -1;
		} else {
			if (cursor == back) {
				cursor = null;
				index = -1;
			}
			back = back.prev;
		}
		length--;
	}
	
	//delete()
	//Deletes the cursor element, making cursor undefined
	//Pre: length()>0, index()>=0
	void delete() {
		if (this.length <= 0 || this.index() < 0)
			throw new RuntimeException("List Error: delete() called on an empty List");
		if (cursor == back)
			deleteBack();
		else if (cursor == front)
			deleteFront();
		else {
			Node N = cursor.next;
			cursor = cursor.prev;
			cursor.next = N;
			index = -1;
			length--;
		}
	}

	// Other Functions ---------------------------------------------------------

	// toString()
	// Overides Object's toString() method.
	public String toString(){
		StringBuffer sb = new StringBuffer();
		Node N = front;
		while (N != null) {
			sb.append(N.toString());
			sb.append(" ");
			N = N.next;
		}
		return new String(sb);
	}
}